suave/backends/redis_transport.go

```diff
@@ -0,0 +1,183 @@
+	}
+)
+
+type RedisPubSubTransport struct {
+	ctx      context.Context
+	cancel   context.CancelFunc
+	redisUri string
+	client   *redis.Client
+}
+
+func NewRedisPubSubTransport(redisUri string) *RedisPubSubTransport {
+	return &RedisPubSubTransport{
+		redisUri: redisUri,
+	}
+}
+
+func (r *RedisPubSubTransport) Start() error {
+	if r.cancel != nil {
+		r.cancel()
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	r.ctx = ctx
+	r.cancel = cancel
+
+	client, err := connectRedis(r.redisUri)
+	if err != nil {
+		return err
+	}
+	r.client = client
+
+	return nil
+}
+
+func (r *RedisPubSubTransport) Stop() error {
+	if r.cancel == nil || r.client == nil {
+		panic("Stop() called before Start()")
+	}
+
+	r.cancel()
+	r.client.Close()
+
+	return nil
+}
+
+func (r *RedisPubSubTransport) Subscribe() (<-chan suave.DAMessage, context.CancelFunc) {
+	ch := make(chan suave.DAMessage, 16)
+	ctx, cancel := context.WithCancel(r.ctx)
+
+	// Each subscriber has its own PubSub as it blocks on receive!
+	pubsub := r.client.Subscribe(ctx, redisUpsertTopic)
+
+	go func() {
+		defer close(ch)
+		defer pubsub.Close()
+
+		for ctx.Err() == nil /* run until Stop() or cancel() called */ {
+			rmsg, err := pubsub.ReceiveMessage(ctx)
+			if err != nil {
+				if ctx.Err() != nil {
+					// Stop() or cancel() called, simply exit
+					log.Info("Redis pubsub: closing subscription")
+					return
+				}
+
+				// Reconnection is not necessary as it's handled by redis.PubSub, simply log the error and continue
+				log.Error("Redis pubsub: error while receiving messages", "err", err)
+				continue
+			}
+
+			var msg suave.DAMessage
+			err = json.Unmarshal([]byte(rmsg.Payload), &msg)
+			if err != nil {
+				log.Trace("Redis pubsub: could not parse message from subscription", "err", err, "msg", rmsg.Payload)
+				continue
+			}
+
+			// For some reason the caller, key, and value fields are not parsed correctly
+			// TODO: debug
+			m := make(map[string]interface{})
+			err = json.Unmarshal([]byte(rmsg.Payload), &m)
+			if err != nil {
+				log.Trace("Redis pubsub: could not parse message from subscription", "err", err, "msg", rmsg.Payload)
+				continue
+			}
+
+			msg.Caller = common.HexToAddress(m["caller"].(string))
+			msg.Key = m["key"].(string)
+			msg.Value = common.FromHex(m["value"].(string))
+
+			log.Debug("Redis pubsub: new message", "msg", msg)
+			select {
+			case <-ctx.Done():
+				log.Info("Redis pubsub: closing subscription")
+				return
+			case ch <- msg:
+				continue
+			default:
+				log.Error("dropping transport message due to channel being blocked")
+				continue
+			}
+		}
+	}()
+
+	return ch, cancel
+}
+
+func (r *RedisPubSubTransport) Publish(message suave.DAMessage) {
+	log.Trace("Rebids pubsub: publishing", "message", message)
+	data, err := json.Marshal(message)
+	if err != nil {
+		panic(fmt.Errorf("could not marshal message: %w", err))
```

Comment:
no panic please

Reply:
On the other hand, under what circumstances would this be the case?
Added a log + early return